### PR TITLE
JWWEB-621: Selectors, that throw

### PR DESCRIPTION
**Pull request by flexxnn**

### CODE REVIEW

File: src/errors/ActiveNetworkNotFound.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// @flow
+
+class ActiveNetworkNotFoundError extends Error {
+  constructor() {
+    const message = 'Active network not found'
```
Comment (stcherenkov) [inline src/errors/ActiveNetworkNotFound.js]: This message is not localized. What use does it have? We could not display it to user.

File: src/errors/tests/ActiveWalletNotFound.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import { DigitalAssetNotFoundError } from '..'
+
+const assetAddress = 'Ethereum'
+
+describe('DigitalAssetNotFoundError', () => {
+  test('exists', () => {
+    expect(DigitalAssetNotFoundError).not.toEqual(undefined)
+  })
+
+  test('can be throwed', () => {
```
Comment (stcherenkov) [inline src/errors/tests/ActiveWalletNotFound.test.js]: Typo, "thrown"

File: src/routes/Upgrade/UpgradeViewContainer.js [hunk 3]
```diff
@@ -75,7 +76,7 @@ function validateMnemonic(bip32XPublicKey: ?string) {
     derivationPath,
   }: UpgradeMnemonicFormFieldValues): UpgradeMnemonicFormFieldErrors => {
     if (!bip32XPublicKey) {
-      throw new Error(t`WalletDataError`)
+      throw new WalletInvalidDataError('', 'bip32XPublicKey is empty')
```
Comment (stcherenkov) [inline src/routes/Upgrade/UpgradeViewContainer.js]: Why is it not `WalletEmptyAddressError` and `WalletEmptyBIP32XPubError`? How do we decide when we use new error class and when we use string description? How do we check errors for specific descriptions?

File: src/store/sagas/balances.js [hunk 4]
```diff
@@ -296,12 +295,8 @@ function* fetchByAssetSuccess(
...
   // check conditions - network, balance, etc
 
-  const digitalAsset: ExtractReturn<typeof selectDigitalAsset> =
-    yield select(selectDigitalAsset, assetAddress)
-
-  if (!digitalAsset) {
-    throw new Error(t`DigitalAssetNotFoundError`)
-  }
+  const digitalAsset: ExtractReturn<typeof selectDigitalAssetOrThrow> =
+    yield select(selectDigitalAssetOrThrow, assetAddress)
```
Comment (stcherenkov) [inline src/store/sagas/balances.js]: Do we throw in sagas? How do sagas recover if error is thrown in the middle of state machine transition?

File: src/errors/ActiveNetworkNotFound.js (hunk 1 above)
Comment (IViol) [reply src/errors/ActiveNetworkNotFound.js]: this message is useless and could be easy removed

File: src/errors/ActiveWalletNotFound.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// @flow
+
+class ActiveWalletNotFoundError extends Error {
+  constructor() {
+    const message = 'Active wallet not found'
```
Comment (IViol) [inline src/errors/ActiveWalletNotFound.js]: this message is useless and could be easy removed

File: src/errors/DigitalAssetNotFound.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// @flow
+
+class DigitalAssetNotFoundError extends Error {
+  address: AssetAddress
+
+  constructor(address: AssetAddress) {
+    const message = `Digital asset with address ${address} not found`
```
Comment (IViol) [inline src/errors/DigitalAssetNotFound.js]: this message is useless and could be easy removed
Also such `const`s better declare in file scope, not in `constructor` or any other method

File: src/errors/WalletInvalidData.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+// @flow
+
+class WalletInvalidDataError extends Error {
```
Comment (IViol) [inline src/errors/WalletInvalidData.js]: better to call it `InconsistentData`
and message is not needed for this type of error

File: src/errors/WalletNotFound.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// @flow
+
+class WalletNotFoundError extends Error {
+  walletId: ?WalletId
+
+  constructor(walletId: ?WalletId) {
+    const message = `Wallet ${walletId || 'null'} not found`
```
Comment (IViol) [inline src/errors/WalletNotFound.js]: ~~message~~

File: src/errors/tests/ActiveNetworkNotFound.test.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
+// @flow
```
Comment (IViol) [inline src/errors/tests/ActiveNetworkNotFound.test.js]: let's move tests to `src/errors`.
now we don't understand how deep can specific `tests` dir could be placed, so it's ok to keep all tests near origin src file

File: src/errors/tests/ActiveWalletNotFound.test.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+// @flow
+
+import { DigitalAssetNotFoundError } from '..'
+
+const assetAddress = 'Ethereum'
```
Comment (IViol) [inline src/errors/tests/ActiveWalletNotFound.test.js]: let's use uppercase for file-scope consts

File: src/routes/Upgrade/UpgradeViewContainer.js (hunk 3 above)
Comment (IViol) [reply src/routes/Upgrade/UpgradeViewContainer.js]: this is called "inconsistency" when we expect to get specific valid data for specific type of wallet, but this data doesn't exist

File: src/store/sagas/balances.js (hunk 4 above)
Comment (IViol) [reply src/store/sagas/balances.js]: good question!
It's ok to `throw` in sagas.
*Service for unhandled exceptions* will care how to handle them after, as I suggested in the confluence doc.

File: src/store/sagas/blocks.js [hunk 11]
```diff
@@ -58,7 +59,7 @@ function* latestBlockSync(networkId: NetworkId): Saga<void> {
         yield select(selectNetworkById, networkId)
 
       if (!network) {
-        throw new Error(t`ActiveNetworkNotFoundError`)
+        throw new ActiveNetworkNotFoundError()
       }
```
Comment (IViol) [inline src/store/sagas/blocks.js]: `selectNetworkByIdOrThrow`?

File: src/store/selectors/networks.js [hunk 12]
```diff
@@ -22,6 +26,16 @@ export function selectCurrentNetworkId(state: AppState): NetworkId {
   return networksPersist.currentNetworkId
 }
 
+export function selectCurrentNetworkIdOrThrow(state: AppState): NetworkId {
+  const networkId = selectCurrentNetworkId(state)
+
+  if (!networkId || networkId === '*') {
```
Comment (IViol) [inline src/store/selectors/networks.js]: why ``networkId === '*'`?

File: src/errors/tests/ActiveNetworkNotFound.test.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+import { ActiveNetworkNotFoundError } from '..'
+
+describe('ActiveNetworkNotFoundError', () => {
+  test('exists', () => {
+    expect(ActiveNetworkNotFoundError).not.toEqual(undefined)
```
Comment (IViol) [inline src/errors/tests/ActiveNetworkNotFound.test.js]: .not.toEqual(undefined) -> toBeDefined()

File: src/errors/tests/ActiveNetworkNotFound.test.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+      expect(err.name).toEqual('ActiveNetworkNotFoundError')
+    }
+  })
+
+  test('can be checked by instanceof', () => {
+    try {
+      throw new ActiveNetworkNotFoundError()
+    } catch (err) {
+      expect(err instanceof ActiveNetworkNotFoundError).toBe(true)
+      expect(err instanceof Error).toBe(true)
```
Comment (IViol) [inline src/errors/tests/ActiveNetworkNotFound.test.js]: `toBeInstanceOf`

File: src/errors/tests/ActiveNetworkNotFound.test.js (hunk 9 above)
Comment (flexxnn) [reply src/errors/tests/ActiveNetworkNotFound.test.js]: A few days ago we discussed that we will use `tests` subfolder for unit tests. 
@stcherenkov @Creepypoke @Dragonesis What do you think about this?

**Comment by stcherenkov** [comment]
Please update errors according to [docs](https://jibrelnetwork.atlassian.net/wiki/spaces/JWALLET/pages/732200961/Concept+of+error+handling). Also use JError as base error class please.

File: src/routes/Upgrade/UpgradeViewContainer.js [hunk 15]
```diff
@@ -36,7 +36,7 @@ import UpgradeView from './UpgradeView'
 function validatePrivateKey(address: ?Address) {
   return ({ privateKey }: UpgradePrivateKeyFormFieldValues): UpgradePrivateKeyFormFieldErrors => {
     if (!address) {
-      throw new WalletInconsistentDataError('Address is empty', '')
+      throw new WalletInconsistentDataError(null, 'Address is empty')
```
Comment (IViol) [inline src/routes/Upgrade/UpgradeViewContainer.js]: `throw new WalletInconsistentDataError('Address is empty')`
why `null` in first param?

File: src/routes/Upgrade/UpgradeViewContainer.js [hunk 16]
```diff
@@ -76,7 +76,7 @@ function validateMnemonic(bip32XPublicKey: ?string) {
     derivationPath,
   }: UpgradeMnemonicFormFieldValues): UpgradeMnemonicFormFieldErrors => {
     if (!bip32XPublicKey) {
-      throw new WalletInconsistentDataError('bip32XPublicKey is empty', '')
+      throw new WalletInconsistentDataError(null, 'bip32XPublicKey is empty')
```
Comment (IViol) [inline src/routes/Upgrade/UpgradeViewContainer.js]: why `null`?

File: src/utils/wallets/getPrivateKey.js [hunk 17]
```diff
@@ -30,7 +30,10 @@ function getPrivateKey(wallet: Wallet, internalKey: Uint8Array, encryptionType:
       !network ||
       !derivationPath
     ) {
-      throw new WalletInconsistentDataError(`Invalid mnemonic type ${type}`, wallet.id)
+      throw new WalletInconsistentDataError(
+        { walletId: wallet.id },
+        `Invalid mnemonic type ${type}`,
```
Comment (IViol) [inline src/utils/wallets/getPrivateKey.js]: `checkMnemonicType(type)` above checks `type === 'mnemonic'`
so message `Invalid mnemonic type ${type}` always will be _"Invalid mnemonic type **mnemonic**"_

**Pull request merged**